Sources/ApolloCodegenLib/ApolloCodegen.swift

```diff
@@ -462,9 +462,24 @@ public class ApolloCodegen {
     default: break
     }
 
-    return try globs.reduce(into: []) { partialResult, glob in
+    var existingPaths: Set<String> = try globs.reduce(into: []) { partialResult, glob in
       partialResult.formUnion(try glob.match())
     }
+    
+    try resolveSymlinksInPaths(&existingPaths)
+    
+    return existingPaths
+  }
+  
+  private static func resolveSymlinksInPaths(_ paths: inout Set<String>) throws {
```

Comment:
This function looks like it's just removing symlink paths, not actually resolving the symlinks. Is that correct?

If so, we should probably change the name of the function. But there is also something awkward about this, because there is implicit knowledge that the paths returned from the glob have _both_ the path with and without the symlink in the `Set`. Is that correct?

I'm feeling like we should be expressing this better and removing the need for implicit knowledge here. Especially since you are now calling `resolvingSymlinksInPath()` inside of the `Glob` code.

If we are going to resolve the symlinks in the Glob, I'm not even clear why we also need this here? I know we had talked about this being the responsibility of the Codegen, not the `Glob` implementation, but it feels like we are violating that anyways now.

Reply:
Good point renaming the function would make since. And yes, the issue currently is that both a real path and symlink path to a file exist in the Set which causes the deletion.

Also the reason `resolvingSymlinksInPath()` is called in `Glob` is specifically due to needing to make sure paths in tests are resolved from `/private/var/....` to just `/var/....`, however that change would not actually fix the non-test issue because the `directoryURL` used at that point in the code would not actually be the symlinked path. The code I showed you earlier where we could do the symlink resolution in `Glob.expand` by building out the paths before adding them to the Set in the return could potentially solve for both the test and non-test issue, will have to check to verify, and if so we may just need to do that since it appears we will need some change inside of `Glob` anyways which is what we were trying to avoid.

Reply:
I don't mind if we do this in the `Glob`, but I feel like we need a clear definition of what we are actually doing.

The real issue here is one of separation of responsibilities.  It's a current behavior that the output of the `Glob` has duplicated paths. That is – two paths that actually point to the same place because of symlinks. If we decide that is an acceptable behavior of `Glob`, because it is just used to pattern match the paths, and there _are_ two paths there technically, then that's okay. Let's document that behavior and then deduplicate the paths in `ApolloCodegen`, as we are doing. In which case the fix for the unit tests `/private/var` needs a different fix probably.

I think it's also reasonable to say that the behavior of the `Glob` struct should do the resolving of symlinks and deduplication of paths, ensuring that its output only points to a single file one time, regardless of how many different paths are symlinked to the file's _actual_ path. If we do that, we can fix the problem in one place, and we should be documenting (and probably testing) that as an intended and defined behavior of our `Glob` matching object.

I don't have a strong opinion on where we do this, only that we do it in a way that defines a responsibility of a single object that is sensible; is documented; is tested; and works in consistently in every case (of course bugs are bugs, so "every case" means "every known case"). 

I think I might be misunderstanding or missing some information here, but having the symlink deduplication in the codegen but also resolving symlinks in the Glob seems to be muddying the single responsibility principle here.

If there is a specific odd edge case where `/private/var` isn't being picked up as a symlink properly we can hard code in a fix for the edge case. But if we are going to resolve all symlinks lets have a clear definition of what that means, where it's happening, and why.

Reply:
I built `Glob` to be a pattern matching engine; it expands given wildcards and outputs matching paths. IMO those should be resolved symlinks and there should _not_ be any duplicates. `Glob` is at the crux of path discovery here and it feels that should be doing the work. I don't think there is ever a need for us to know both the symlink and actual path?